https://twistedmatrix.com/trac/ticket/8858

glyph opened the pull request.

glyph committed:
add traceback to formatEventAsClassicLogText

rather than (incorrectly) in FileLogObserver, where it can mess up
structured log output, such as JSON

glyph committed:
Gross Violation Of Specifications

wsanchez commented on src/twisted/logger/test/test_json.py:
```diff
@@ -280,6 +280,23 @@ def test_observeWritesEmptyRecordSeparator(self):
+            1 / 0
+        except:
+            logger.failure("failed as expected")
+        print("AN_VALUE", io.getvalue())
```
remove print()

wsanchez commented on src/twisted/logger/test/test_json.py:
```diff
@@ -280,6 +280,23 @@ def test_observeWritesEmptyRecordSeparator(self):
+        reader = StringIO(io.getvalue())
+        events = list(eventsFromJSONLogFile(reader))
+        self.assertEqual(len(events), 1)
+
```
This probably should check that the de-serialized event contains a `log_failure` key?

wsanchez replied:
Also wondering whether we need to ensure that the failure's traceback is being captured in the JSON.

Julian replied:
(Agree on both of those)

glyph committed:
Merge branch 'trunk' into 8858-tracebacks

glyph committed:
ITS

glyph committed:
compare less strictly

glyph committed:
print() removed as per wsanchez's comment

glyph committed:
assert some stuff about traceback structure as per review feedback

glyph commented:
The Windows build failure is logged here: https://twistedmatrix.com/trac/ticket/8870

markrwilliams commented on src/twisted/logger/test/test_json.py:
```diff
@@ -280,6 +280,40 @@ def test_observeWritesEmptyRecordSeparator(self):
+        [event] = events
+        [beforeEvent] = before
+        self.assertIn("log_failure", event)
+        f = event["log_failure"]
```
I'd prefer longer variable names for the Failures here.

glyph replied:
Sure.

markrwilliams commented on src/twisted/logger/test/test_json.py:
```diff
@@ -280,6 +280,40 @@ def test_observeWritesEmptyRecordSeparator(self):
+        deserializedTraceback = f.getTracebackObject()
+        previousTraceback = b.getTracebackObject()
+        def checkOneTraceback(tb):
+            self.assertEqual(tb.tb_frame.f_code.co_filename.rstrip("co"),
```
An additional check that confirms the `f_code.co_name` is the same as the test method's would be nice.  I guess you didn't include that, though, because it would make the code depend on the method's name?

glyph replied:
Yeah. If one detail is accurate, that's close enough; this is not a test for every detail of traceback generation or serialization.

markrwilliams commented on src/twisted/topfiles/8858.bugfix:
```diff
@@ -0,0 +1 @@
+twisted.logger.jsonFileLogObserver no longer emits non-JSON data into its file in the case of tracebacks; additionally, twisted.logger.formatEventAsClassicLogText now includes traceback information for the log event it is passing.
```
twisted.logger.jsonFileLogObserver no longer emits non-JSON ~~data into its file in the case of~~ **tracebacks into its file**; additionally, twisted.logger.formatEventAsClassicLogText now includes traceback information for the log event it ~~is passing~~ **formats**.

glyph replied:
Yes, better grammar.

markrwilliams reviewed (commented):
Minor, non-merge-blocking comments.

glyph committed:
better topfile, as suggested in review

glyph committed:
avoid short names (as per review)

do this by naming fewer thing and factoring more of the common
assertions into the function
